### ISSUE TITLE
NextTopLoader stopped working on Nextjs 14.0.3

### ISSUE THREAD

**Issue by ajsystem** [TheSGJ/nextjs-toploader#56]
Hello, just to report that on a project with the last Nextjs 14.0.3, the toploader keeps loading and never finish or completes.

But there aren't errors on console server/client.

As a workaround to get it working install 14.0.2

**Comment by rrojan** [comment]
Same here. the progress bar is stuck at the end and never finishes loading since NextJS 14.0.3.

It also looks less snappy, if that makes sense. At the time where it previously completed loading, it now goes to around at 60-70% progress and slowly crawls its way to the end, but doesn't stop loading.

**Comment by SuhelMakkad** [comment]
+1

**Comment by jonathanwilke** [comment]
Same problem for me. Downgrading to 14.0.2 fixed it.

**Comment by TheSGJ** [comment]
Next js 14.0.3 added experimental support for history.pushState and history.replaceState. That's why it's messing with history.pushState() method in the code.
Here's the pr which introduced it:
vercel/next.js/pull/58335

**Comment by putramaghfirah** [comment]
same problem.

**Comment by croofec** [comment]
+1

**Comment by sengawasenga** [comment]
I got the same problem, how can we solve it ?

**Comment by omarqra** [comment]
+1

**Comment by PreetiYdv** [comment]
same issue.

**Comment by Deri-Kurniawan** [comment]
same here

**Comment by PedroL22** [comment]
same here

**Comment by besufkadmenji** [comment]
Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
`
import * as NProgress from "nprogress";

  useEffect(() => {
    NProgress.done();
  }, [pathname, router]);
`

**Comment by Jings** [comment]
> Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
> 
> ```
> import * as NProgress from "nprogress";
> 
> useEffect(() => {
>     NProgress.done();
> }, [pathname, router]);
> ```

This works for now.
Hopefully the proper fix will come soon :)

**Comment by joshke** [comment]
+1

**Comment by Super-Kenil** [comment]
+1

**Comment by netstermonir** [comment]
> Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
> 
> ```
> import * as NProgress from "nprogress";
> 
> useEffect(() => {
>     NProgress.done();
> }, [pathname, router]);
> ```

why root layout client-side rendering? I hope it is not good practiche

**Comment by Jings** [comment]
> > Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
> > ```
> > import * as NProgress from "nprogress";
> > 
> > useEffect(() => {
> >     NProgress.done();
> > }, [pathname, router]);
> > ```
> 
> why root layout client-side rendering? I hope it is not good practiche

@besufkadmenji is only talking about your top most client component not root layout client side rendering. If this is your root layout, fine. For me it is not but instead two or three pages that are the top most client components.

**Comment by vnevermore** [comment]
same problem here

**Comment by gokulkrishh** [comment]
> Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
> 
> ```
> import * as NProgress from "nprogress";
> 
> useEffect(() => {
>     NProgress.done();
> }, [pathname, router]);
> ```

Yep this worked for me. But we need proper fix. Thanks for suggesting.

**Comment by NestedGateway** [comment]
Also facing problem with nextjs 14. I am still using it because there is no such alternative to it. All other all manual.

**Comment by gustaveWPM** [comment]
> Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
> 
> ```
> import * as NProgress from "nprogress";
> 
> useEffect(() => {
>     NProgress.done();
> }, [pathname, router]);
> ```

Thank you so much!
I added it at the top of my `<Providers>` component.

My root layout looks like this:
```tsx
    <html lang={language} dir={dir}>
      <body>
        <div>
          <Providers locale={language}>
            <NextTopLoader {...} />
            {children}
// ...
```

So that's a really good workaround, imo.

**Comment by andrewmumblebee** [comment]
Potentially will be fixed by https://github.com/vercel/next.js/pull/58861/files

**Comment by nunesunil** [comment]
Fixed by https://github.com/vercel/next.js/pull/58861  and released in [v14.0.4-canary.37](https://github.com/vercel/next.js/commit/141dbb2663c2f7605df9a19af2ed5154e7b6c881).

**Comment by Super-Kenil** [comment]
@ajsystem , I checked the latest Nextjs v14.0.4. and nextjs-toploader works fine in that version. I think this issues should be closed now

**Comment by ImamAlfariziSyahputra** [comment]
> @ajsystem , I checked the latest Nextjs v14.0.4. and nextjs-toploader works fine in that version. I think this issues should be closed now

let see what other people opinions first, just to make sure before it's closed😉

**Comment by gokulkrishh** [comment]
@ImamAlfariziSyahputra @Super-Kenil Yep issue is fixed in v14.0.4 for my app [expense.fyi](expense.fyi)

**Comment by Super-Kenil** [comment]
@ImamAlfariziSyahputra . I had already tested it in 3 of my projects, then I commented here.
@gokulkrishh thanks for confirming

**Comment by gustaveWPM** [comment]
Fixed!

**Comment by bschwartz10** [comment]
Still experiencing the stuck loader when its 99% done.
```
"next": "^14.0.4"
"nextjs-toploader": "^1.6.4"
"nprogress": "^0.2.0"
```

**Comment by lkk270** [comment]
> Still experiencing the stuck loader when its 99% done. This is happening specifically when using the custom `useRouter` solution from this issue: [#10 (comment)](https://github.com/TheSGJ/nextjs-toploader/issues/10#issuecomment-1809417610)
> 
> ```
> "next": "^14.0.4"
> "nextjs-toploader": "^1.6.4"
> "nprogress": "^0.2.0"
> ```

I have the same packages and versions installed, but for me, the top loader never even shows. I have `<NextTopLoader />` within the` <body> `of the root `layout.tsx` file.

**Comment by B33fb0n3** [comment]
Still not working :(
```
    "next": "14.1.0",
    "nextjs-toploader": "^1.6.4",
```

**Comment by sho-pb** [comment]
Here are my solutions.🥹

```ts
import { NavigateOptions } from 'next/dist/shared/lib/app-router-context.shared-runtime';
import { useRouter as useNextRouter, usePathname } from 'next/navigation';
import { useCallback, useEffect, useRef, useState } from 'react';
import NProgress from 'nprogress';

export const useRouter = () => {
  const router = useNextRouter();
  const pathname = usePathname();
  const prevPathnameRef = useRef(pathname);
  const [isLoading, setIsLoading] = useState(false);

  const replace = useCallback(
    (href: string, options?: NavigateOptions | undefined) => {
      href !== pathname && NProgress.start();
      setIsLoading(true);
      prevPathnameRef.current = pathname;

      router.replace(href, options);
    },
    [router, pathname],
  );

  const push = useCallback(
    (href: string, options?: NavigateOptions | undefined) => {
      href !== pathname && NProgress.start();
      setIsLoading(true);
      prevPathnameRef.current = pathname;

      router.push(href, options);
    },
    [router, pathname],
  );

  useEffect(() => {
    if (isLoading && pathname !== prevPathnameRef.current) {
      NProgress.done();
      setIsLoading(false);
    }
  }, [pathname, isLoading]);

  return {
    ...router,
    replace,
    push,
  };
};

```

**Comment by sho-pb** [comment]
my solutions

```tsx
// NextTopLoader.tsx
'use client';

import Loader from 'nextjs-toploader';
import { usePathname } from 'next/navigation';

const NextTopLoader = () => {
  const pathname = usePathname();

  useEffect(() => {
    NProgress.done();
  }, [pathname]);

  return (
    <Loader />
  )
}
```

```tsx
// useRouter.ts
import { NavigateOptions } from 'next/dist/shared/lib/app-router-context.shared-runtime';
import { useRouter as useNextRouter, usePathname } from 'next/navigation';
import { useCallback } from 'react';
import NProgress from 'nprogress';

export const useRouter = () => {
  const router = useNextRouter();
  const pathname = usePathname();

  const replace = useCallback(
    (href: string, options?: NavigateOptions) => {
      href !== pathname && NProgress.start();
      router.replace(href, options);
    },
    [router, pathname],
  );

  const push = useCallback(
    (href: string, options?: NavigateOptions) => {
      href !== pathname && NProgress.start();
      router.push(href, options);
    },
    [router, pathname],
  );

  return {
    ...router,
    replace,
    push,
  };
};
```

**Comment by Super-Kenil** [comment]
@B33fb0n3 

> Still not working :(
> 
> ```
> "next": "14.1.0",
> "nextjs-toploader": "^1.6.4",
> ```

I am using nextjs-toploader, With same version you're using, Both **Next** and **nextjs-toploader** and it seems to be working perfectly on my end. 

For fixing it, make sure you are not using it multiple times unnecessarily in your App. 
If it doesn't get fixed, please create a Reproduction so that we could take a look at it

**Comment by Ibinola** [comment]
> Possible fix: run NProgress.done() inside your top level client component, this will cleanup any stuck loader
> 
> ```
> import * as NProgress from "nprogress";
> 
> useEffect(() => {
>     NProgress.done();
> }, [pathname, router]);
> ```

What file am i adding this ?

**Comment by moon2850088** [comment]
> my solutions
> 
> ```tsx
> // NextTopLoader.tsx
> 'use client';
> 
> import Loader from 'nextjs-toploader';
> import { usePathname } from 'next/navigation';
> 
> const NextTopLoader = () => {
>   const pathname = usePathname();
> 
>   useEffect(() => {
>     NProgress.done();
>   }, [pathname]);
> 
>   return (
>     <Loader />
>   )
> }
> ```
> 
> ```tsx
> // useRouter.ts
> import { NavigateOptions } from 'next/dist/shared/lib/app-router-context.shared-runtime';
> import { useRouter as useNextRouter, usePathname } from 'next/navigation';
> import { useCallback } from 'react';
> import NProgress from 'nprogress';
> 
> export const useRouter = () => {
>   const router = useNextRouter();
>   const pathname = usePathname();
> 
>   const replace = useCallback(
>     (href: string, options?: NavigateOptions) => {
>       href !== pathname && NProgress.start();
>       router.replace(href, options);
>     },
>     [router, pathname],
>   );
> 
>   const push = useCallback(
>     (href: string, options?: NavigateOptions) => {
>       href !== pathname && NProgress.start();
>       router.push(href, options);
>     },
>     [router, pathname],
>   );
> 
>   return {
>     ...router,
>     replace,
>     push,
>   };
> };
> ```

It works for me. Thanks for your solution !

**Comment by rohankm** [comment]
> my solutions
> 
> ```tsx
> // NextTopLoader.tsx
> 'use client';
> 
> import Loader from 'nextjs-toploader';
> import { usePathname } from 'next/navigation';
> 
> const NextTopLoader = () => {
>   const pathname = usePathname();
> 
>   useEffect(() => {
>     NProgress.done();
>   }, [pathname]);
> 
>   return (
>     <Loader />
>   )
> }
> ```
> 
>  ```tsx
> // useRouter.ts
> import { NavigateOptions } from 'next/dist/shared/lib/app-router-context.shared-runtime';
> import { useRouter as useNextRouter, usePathname } from 'next/navigation';
> import { useCallback } from 'react';
> import NProgress from 'nprogress';
> 
> export const useRouter = () => {
>   const router = useNextRouter();
>   const pathname = usePathname();
> 
>   const replace = useCallback(
>     (href: string, options?: NavigateOptions) => {
>       href !== pathname && NProgress.start();
>       router.replace(href, options);
>     },
>     [router, pathname],
>   );
> 
>   const push = useCallback(
>     (href: string, options?: NavigateOptions) => {
>       href !== pathname && NProgress.start();
>       router.push(href, options);
>     },
>     [router, pathname],
>   );
> 
>   return {
>     ...router,
>     replace,
>     push,
>   };
> };
> ```

this works for the latest nextjs

**Comment by Kamleshpaul** [comment]
> my solutions
> 
> ```tsx
> // NextTopLoader.tsx
> 'use client';
> 
> import Loader from 'nextjs-toploader';
> import { usePathname } from 'next/navigation';
> 
> const NextTopLoader = () => {
>   const pathname = usePathname();
> 
>   useEffect(() => {
>     NProgress.done();
>   }, [pathname]);
> 
>   return (
>     <Loader />
>   )
> }
> ```
> 
> ```tsx
> // useRouter.ts
> import { NavigateOptions } from 'next/dist/shared/lib/app-router-context.shared-runtime';
> import { useRouter as useNextRouter, usePathname } from 'next/navigation';
> import { useCallback } from 'react';
> import NProgress from 'nprogress';
> 
> export const useRouter = () => {
>   const router = useNextRouter();
>   const pathname = usePathname();
> 
>   const replace = useCallback(
>     (href: string, options?: NavigateOptions) => {
>       href !== pathname && NProgress.start();
>       router.replace(href, options);
>     },
>     [router, pathname],
>   );
> 
>   const push = useCallback(
>     (href: string, options?: NavigateOptions) => {
>       href !== pathname && NProgress.start();
>       router.push(href, options);
>     },
>     [router, pathname],
>   );
> 
>   return {
>     ...router,
>     replace,
>     push,
>   };
> };
> ```

this works

**Comment by Emiliano-Bucci** [comment]
> ```tsx
> back
> ```

@sho-pb Hi! Is this also working with the browser back button? :)
